core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java

```diff
@@ -312,25 +312,30 @@ public int hashCode() {
         return nodeId.hashCode();
     }
 
+    private String toString;
```

Comment:
OK :) here is what I think:

1) Every discovery node will logged at least once (due to the log lines you mentioned).
2) Some nodes will be converted multiple times at special occasions (recovery source /targets , cluster state publishing logs the master node again and again).

However, we have to take into account that the above isn't neccesarily done on the same discovery node objects. The vast majority of node *objects* will never have the toString called on them more than once (and likely not even once, since DiscoveryNode(s) objects will be [replaced])(https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/AbstractDiffable.java#L37) on every change.

So considering this I don't think we need to pre-cache the toString  value. I also have my doubts about the cache-on-demand solution and whether it's worth the complexity but was willing to go along.